File: backend/api/services/interpreter_matching.py
"""Language-pair routing for interpreter dispatch and session acceptance."""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ..models import Profile, Session


def interpreter_speaks_language(languages: list | None, language: str) -> bool:
    if not languages or not language:
        return False
    return language in languages


def interpreter_supports_language_pair(
    languages: list | None,
    language_from: str,
    language_to: str,
) -> bool:
    """Interpreter must speak both sides of the requested pair."""
    if not languages:
        return False
    return interpreter_speaks_language(languages, language_from) and interpreter_speaks_language(
        languages, language_to
    )


def is_incoming_call_for_interpreter(
    session: Session,
    interpreter_external_id: str,
    interpreter_languages: list | None,
) -> bool:
    if session.status != "incoming":
        return False

    if session.interpreter_id:
        profile = getattr(session.interpreter, "profile", None)
        assigned_id = profile.external_id if profile else None
        return assigned_id == interpreter_external_id

    if session.scheduled_time != "instant":
        return False

    return interpreter_supports_language_pair(
        interpreter_languages,
        session.language_from,
        session.language_to,
    )


def can_interpreter_accept_session(session: Session, profile: Profile) -> bool:
    if session.status != "incoming":
        return False

    if profile.role != "interpreter" or profile.status != "active":
        return False

    if session.interpreter_id:
        assigned = getattr(session.interpreter, "profile", None)
        if assigned and assigned.external_id != profile.external_id:
            return False

    return interpreter_supports_language_pair(
        profile.languages,
        session.language_from,
        session.language_to,
    )

File: frontend/src/lib/interpreterMatching.ts
import { Session } from "../types";

/** Interpreter must list the language in their profile. */
export function interpreterSpeaksLanguage(
  languages: string[] | undefined,
  language: string
): boolean {
  if (!languages?.length || !language) return false;
  return languages.includes(language);
}

/** Interpreter must speak both sides of the requested pair. */
export function interpreterSupportsLanguagePair(
  languages: string[] | undefined,
  languageFrom: string,
  languageTo: string
): boolean {
  if (!languages?.length) return false;
  return (
    interpreterSpeaksLanguage(languages, languageFrom) &&
    interpreterSpeaksLanguage(languages, languageTo)
  );
}

/** Whether an incoming session should ring on this interpreter's desk. */
export function isIncomingCallForInterpreter(
  session: Pick<Session, "status" | "interpreterId" | "scheduledTime" | "languageFrom" | "languageTo">,
  interpreterId: string,
  interpreterLanguages: string[] | undefined
): boolean {
  if (session.status !== "incoming") return false;

  if (session.interpreterId) {
    return session.interpreterId === interpreterId;
  }

  if (session.scheduledTime !== "instant") return false;

  return interpreterSupportsLanguagePair(
    interpreterLanguages,
    session.languageFrom,
    session.languageTo
  );
}

/** First broadcast or direct-dial call this interpreter is eligible to accept. */
export function findIncomingSessionForInterpreter(
  sessions: Session[],
  interpreterId: string,
  interpreterLanguages: string[] | undefined,
  dismissedSessionId?: string | null
): Session | undefined {
  return sessions.find(
    (session) =>
      session.id !== dismissedSessionId &&
      isIncomingCallForInterpreter(session, interpreterId, interpreterLanguages)
  );
}

export function isDirectDialSession(session: Pick<Session, "interpreterId">, interpreterId: string): boolean {
  return Boolean(session.interpreterId && session.interpreterId === interpreterId);
}
